js/stores/simulation-store.ts

```diff
@@ -734,20 +737,53 @@ export class SimulationStore {
   }
 
   // Helpers.
+  getDataSamplesDataset(): IDataset {
+    return {
+      type: "dataset",
+      version: 1,
+      properties: DATASET_PROPS,
+      rows: this.dataSamples.map(sample =>
+        // Type casting is necessary, as some of the sample types are not basic. But they should not be added to
+        // dataset props list anyway.
+        DATASET_PROPS.map(propName => sample[propName]) as (string | number)[]
+      )
+    };
+  }
+
   saveInteractiveState() {
-    setInteractiveState<IInteractiveState>({
+    const dataset = this.getDataSamplesDataset();
+    const newState: IInteractiveState = {
+      ...this.interactiveState,
       answerType: "interactive_state",
-      dataset: {
-        type: "dataset",
-        version: 1,
-        properties: DATASET_PROPS,
-        rows: this.dataSamples.map(sample =>
-          // Type casting is necessary, as some of the sample types are not basic. But they should not be added to
-          // dataset props list anyway.
-          DATASET_PROPS.map(propName => sample[propName]) as (string | number)[]
-        )
+      dataset,
+    };
+    setInteractiveState<IInteractiveState>(newState);
+
+    // Delay snapshot taking, so there's time for the view to re-render itself after state change.
+    // Another approach could be to take snapshots directly in the components after they re-render themselves and pass
+    // them to the simulation store. But it'll spread the logic across multiple components and might be more complex.
+    setTimeout(() => {
+      // This is client side timestamp and it's possible that the clock is set incorrectly. However, this is perfectly
+      // fine here, as it's only used to compare the order of requests.
+      const requestTimestamp = Date.now();
+      const snapshotPromises = [takePlanetViewSnapshot()];
+      if (this.crossSectionVisible) {
+        snapshotPromises.push(takeCrossSectionSnapshot());
       }
-    });
+      Promise.all(snapshotPromises).then(([planetViewSnapshot, crossSectionSnapshot]: string[]) => {
+        if (requestTimestamp < this.lastSnapshotRequestTimestamp) {
```

Comment:
Since `takeSnapshot` can reject I think we need a `.catch()` here that just sets `newState` and skips the snapshots.

Reply:
Note that newState is set even before we try to take snapshots. Line 760 above. So, in most cases, we'll update the state twice, but that minimizes the risk of losing data. Sample data will be saved immediately, and we'll wait only for snapshots. That's why we don't really need to `catch` in this case.

Also, next PR adds `.catch` anyway:
https://github.com/concord-consortium/tectonic-explorer/pull/291